File: src/main/java/com/teamwizardry/wizardry/common/item/ItemStaff.java
package com.teamwizardry.wizardry.common.item;

import org.jetbrains.annotations.NotNull;

import com.teamwizardry.wizardry.common.spell.SpellCompiler;
import com.teamwizardry.wizardry.common.spell.component.Interactor;

import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.util.Hand;
import net.minecraft.util.TypedActionResult;
import net.minecraft.world.World;

public class ItemStaff extends Item implements INacreProduct.INacreDecayProduct {
    public ItemStaff(Settings settings) {
        super(settings);
        // TODO - find where property overrides are
//        this.addPropertyOverride(new Identifier("staff_state"),
//                (stack, world, entity) -> stack.hasTag() && stack.getTag().contains(StringConsts.SPELL_DATA) ? 1 : 0);
    }

    @Override
    public @NotNull TypedActionResult<ItemStack> use(World world, PlayerEntity player, Hand hand) {
        ItemStack itemstack = player.getStackInHand(hand);

        // TODO: Test spell, delete when spell crafting is finished
        if (!world.isClient) {
            Interactor caster = new Interactor(player);
            SpellCompiler.get()
                    .compileSpell(new ItemStack(Items.BEEF),
                            new ItemStack(Items.LEATHER),
                            new ItemStack(Items.LAPIS_LAZULI),
                            new ItemStack(Items.LAPIS_LAZULI),
                            new ItemStack(Items.MAGMA_CREAM))
                    .toInstance(caster)
                    .run(world, caster);
        }

        return TypedActionResult.success(itemstack);
    }
}

File: src/main/java/com/teamwizardry/wizardry/common/item/ItemStaff.kt
package com.teamwizardry.wizardry.common.item

import net.minecraft.item.*
import net.minecraft.util.Hand

class ItemStaff(settings: Settings?) : Item(settings), INacreDecayProduct {
    override fun use(world: World, player: PlayerEntity, hand: Hand): TypedActionResult<ItemStack> {
        val itemstack: ItemStack = player.getStackInHand(hand)

        // TODO: Test spell, delete when spell crafting is finished
        if (!world.isClient) {
            val caster = Interactor(player)
            SpellCompiler.Companion.get()
                .compileSpell(
                    ItemStack(Items.BEEF),
                    ItemStack(Items.LEATHER),
                    ItemStack(Items.LAPIS_LAZULI),
                    ItemStack(Items.LAPIS_LAZULI),
                    ItemStack(Items.MAGMA_CREAM)
                )
                .toInstance(caster)
                .run(world, caster)
        }
        return TypedActionResult.success<ItemStack>(itemstack)
    }
}